File: leetcode_src/src/MyCalendarI.java
import java.util.Set;
import java.util.TreeSet;

class MyCalendarI {
    TreeSet<int[]> booked;

    public MyCalendarI() {
        booked = new TreeSet<>((a, b) -> a[0] - b[0]);
    }

    public boolean book(int start, int end) {
        int[] meeting = new int[]{start, end};
        if(booked.isEmpty()) {
            booked.add(meeting);
            return true;
        }
        int[] tmp = new int[]{end, 0};
        int[] arr = booked.ceiling(tmp);
        int[] prev = (arr == null) ? booked.last() : booked.lower(tmp);
        if(prev == null || start >= prev[1]) {
            booked.add(meeting);
            return true;
        }
        return false;
    }
}

File: leetcode_src/bin/MyCalendarI.py
import bisect
from typing import *

class MyCalendar:
    def __init__(self):
        self.booked = []

    def book(self, start: int, end: int) -> bool:
        i = bisect.bisect_right(self.booked, start)
        if i % 1 or (i < len(self.booked) and self.booked[i] < end):
            return False
        self.booked[i:i] = [start, end]
        return True


if __name__ == '__main__':
    cal = MyCalendar()
    cal.book(1, 2)
    cal.book(3, 4)
    cal.book(2, 3)
